### ISSUE TITLE
Replace: could not replace all placeholders in SecretTemplate data: [replaceString: missing Vault value for placeholder passcode in string password: <passcode>]

### ISSUE THREAD

**Issue by s-u-b-h-a-k-a-r** [IBM/argocd-vault-plugin#42]
**Describe the bug**
Eventhough the vault value exists in vault it throws error that value doesnt exist
**To Reproduce**
**Create vault.json with vault details.**
{
  "VAULT_ADDR": "http://172.18.0.2:31030",
  "AUTH_TYPE": "approle",
  "TYPE": "vault",
  "ROLE_ID": "ee948507-e4b5-1982-5186-d1de6db3e946",
  "SECRET_ID": "fdde601e-fdb8-0ddb-050a-34ef1657a447"
}

**Create secret.yaml with place holders i.e with angular brackets <>**
kind: Secret
apiVersion: v1
metadata:
  name: example-secret
  annotations:
    avp_path: "secret/data/mysql/webapp/"
type: Opaque
data:
  **

> password: "<passcode>"

**

Run the below command:

**argocd-vault-plugin generate secret.yaml -c vault.json**
Error: Replace: could not replace all placeholders in SecretTemplate data: [replaceString: missing Vault value for placeholder passcode in string password: <passcode>]
Usage:
  argocd-vault generate <path> [flags]

Flags:
  -c, --config-path string   path to a file containing Vault configuration (YAML, JSON, envfile) to use
  -h, --help                 help for generate
  -s, --secret-name string   name of a Kubernetes Secret containing Vault configuration data in the argocd namespace of your ArgoCD host (Only available when used in ArgoCD)

Replace: could not replace all placeholders in SecretTemplate data: [replaceString: missing Vault value for placeholder passcode in string password: <passcode>]
**Login to vault and read the key**
 $ vault kv get secret/mysql/webapp
====== Metadata ======
Key              Value
---              -----
created_time     2021-01-30T08:00:26.757330401Z
deletion_time    n/a
destroyed        false
version          1

====== Data ======
Key         Value
---         -----
**passcode    my-long-passcode**

**Comment by werne2j** [comment]
@s-u-b-h-a-k-a-r this error is thrown `replaceString: missing Vault value for placeholder passcode in string password: ` if there is no placeholder to match the value in vault. It is hard to tell from your issue but it looks like your yaml does not have the correct style of placeholder. It should look like
```
kind: Secret
apiVersion: v1
metadata:
  name: example-secret
  annotations:
    avp_path: "secret/data/mysql/webapp/"
type: Opaque
data:
  password: <passcode> 
```
The value of the vault key should go inside the placeholder syntax

**Comment by jkayani** [comment]
Also,

>  vault kv get secret/mysql/webapp

is what you ran to print out `passcode my-long-passcode`

but in the `secret.yaml`, you defined the path for the plugin to look in was:

> "secret/data/mysql/webapp/"

I wonder if you removed that `data` if it might work

**Comment by tico24** [comment]
I believe I have had this issue. The problem lies with the plugin not able to put a non-base64-encoded string into a secret.

The workaround is to change your second to last line from "data:" to "stringData:". Kubernetes will then allow non-encoded data into the secret.

The error message is really confusing and this had me stumped for a while.

**Comment by tico24** [comment]
> I wonder if you removed that `data` if it might work

@jkayani I believe that /data/ is required (depending on your vault setup). I have discussed this in #47 if you're interested.

**Comment by werne2j** [comment]
@tico24 @s-u-b-h-a-k-a-r the issue is with our handling of kv-v2. We have identified the problem and are working on a fix that should solve this issue

**Comment by jkayani** [comment]
Great catch @tico24 thanks, didn't realize that

**Comment by werne2j** [comment]
A fix has been merged into main and we be included in the 0.3.0 release

**Comment by werne2j** [comment]
Released in 0.3.0 https://github.com/IBM/argocd-vault-plugin/releases/tag/v0.3.0
